packages/hierarchies-react/src/presentation-hierarchies-react/internal/TreeLoader.ts

```diff
@@ -61,56 +67,54 @@ export class TreeLoader implements ITreeLoader {
     ).pipe(
       toArray(),
       catchError((err) => {
-        const nodeProps = {
-          id: `${infoNodeIdBase}-Unknown`,
-          parentId: parent.id,
-        };
         let hierarchyLoadErrorType: "unknown" | "timeout" = "unknown";
         if (err instanceof Error) {
-          nodeProps.id = `${infoNodeIdBase}-${err.message}`;
           if (isRowsLimitError(err)) {
             this._onHierarchyLimitExceeded({ parentId: parent.id, filter: instanceFilter, limit: err.limit });
-            return of([{ ...nodeProps, type: "ResultSetTooLarge" as const, resultSetSizeLimit: err.limit }]);
+            return of({ id: `${infoNodeIdBase}-${err.message}`, type: "ResultSetTooLarge" as const, resultSetSizeLimit: err.limit });
           }
           if (isTimeoutError(err)) {
             hierarchyLoadErrorType = "timeout";
           }
         }
 
         this._onHierarchyLoadError({ parentId: parent.id, type: hierarchyLoadErrorType, error: err });
-        return of([
-          {
-            ...nodeProps,
-            type: "Unknown" as const,
-            message: "Failed to create hierarchy level",
-          },
-        ]);
+        return of({
+          id: `${infoNodeIdBase}-Unknown`,
+          type: "Unknown" as const,
+          message: "Failed to create hierarchy level",
+        });
       }),
-      map(
-        (childNodes): LoadedTreePart => ({
+      map((childNodes): LoadedTreePart => {
+        if ("length" in childNodes) {
+          return instanceFilter && childNodes.length === 0
+            ? {
+                parentId: parent.id,
+                error: {
+                  id: `${infoNodeIdBase}-no-filter-matches`,
+                  type: "NoFilterMatches" as const,
+                },
+              }
+            : {
+                parentId: parent.id,
+                loadedNodes: childNodes.map(treeModelNodesFactory),
+              };
+        }
+        return {
           parentId: parent.id,
-          loadedNodes:
-            instanceFilter && childNodes.length === 0
-              ? [
-                  {
-                    id: `${infoNodeIdBase}-no-filter-matches`,
-                    parentId: parent.id,
-                    type: "NoFilterMatches" as const,
-                  },
-                ]
-              : childNodes.map(treeModelNodesFactory),
-        }),
-      ),
+          error: childNodes,
```

Comment:
Here's what I see:
1. `toArray` gives us either an array of nodes or throws an error.
2. `catchError` gets an error and maps it to `ErrorInfo` object. After the operator we may get either an array of nodes or an `ErrorInfo`.
3. `map` has to do some interesting checks to determine what it got from `catchError`, then returns a `LoadedTreePart`, possibly with error state even if an error wasn't thrown.

Here's what I suggest to simplify this:
1. `toArray` gives us either an array of nodes or throws an error.
2. `map` gets an array, does the filter + nodes count and throws an error if needed, or returns a `LoadedTreePart` with `loadedNodes`.
3. `catchError` gets an error, maps it to `ErrorInfo` object, and returns a `LoadedTreePart` with `error`.

What do you think?